File: functions/presence.js
const {getDatabase} = require("firebase-admin/database");
const {getFirestore, FieldValue} = require("firebase-admin/firestore");
const {MODES, STALE_ENTRY_MS} = require("./matchmaking");
const {readEventWindowConfig, upcomingWindowDayKey} = require("./eventWindow");

/**
 * How many people are actually here right now.
 *
 * This is the highest-leverage piece of the Sixes and Sevens window, and
 * the reason is worth stating plainly: the objection to showing up is
 * almost never "I don't want to roast someone." It is "there won't be
 * anyone there." A real number answers that objection directly, and during
 * the window it answers it in the affirmative, which is the entire point of
 * concentrating everyone into one hour.
 *
 * It counts people WAITING TO BATTLE (in a matchmaking queue) and people
 * ALREADY IN ONE (paired, match not yet settled). Both are "here to
 * battle"; someone idly browsing the leaderboard is not, and counting them
 * would inflate the number into a lie. The number has to stay honest or it
 * stops doing its job the first time someone queues against it and finds
 * nobody.
 */

/** Entries this old were abandoned by a crashed or closed client and must
 * not be counted - a ghost entry inflating the number is exactly the kind
 * of dishonesty this feature cannot afford. Same threshold the pairing
 * logic uses to refuse to pair against them. */
const PRESENCE_STALE_MS = STALE_ENTRY_MS;

/**
 * Counts live queue entries out of a raw per-mode queue snapshot.
 *
 * Pure, so the counting rules can be tested without a database.
 */
function countQueue(queuesByMode, nowMs, staleMs = PRESENCE_STALE_MS) {
  let waiting = 0;
  let matched = 0;
  for (const queue of Object.values(queuesByMode || {})) {
    for (const entry of Object.values(queue || {})) {
      if (!entry || typeof entry !== "object") continue;
      const joinedAt = Number(entry.joinedAt) || 0;
      if (nowMs - joinedAt > staleMs) continue;
      if (entry.status === "matched") matched += 1;
      else if (entry.status === "waiting") waiting += 1;
    }
  }
  return {waiting, matched, total: waiting + matched};
}

/**
 * Reads the queues and publishes the count to `stats/presence`.
 *
 * WHY A SCHEDULED JOB rather than a callable clients poll, or a write on
 * every pairing poll:
 * - A callable costs one invocation per client per poll, so cost scales
 *   with users watching the number - the opposite of what's wanted during
 *   the traffic spike this feature exists to create.
 * - Writing from pollMatchmaking would fire every 3 seconds per waiting
 *   client. Twenty people queueing would be roughly 576k Firestore writes a
 *   day against a 20k/day free tier.
 * - A fixed once-a-minute job costs 1,440 writes a day whatever the user
 *   count, and every client reads it through one cheap snapshot listener.
 *
 * It also has no staleness hole: if everyone closes the app, the next tick
 * still publishes zero, whereas an activity-driven write would leave the
 * last non-zero number frozen on screen with nobody around.
 */
async function publishOnlineCount() {
  const db = getDatabase();
  const queuesByMode = {};
  for (const mode of MODES) {
    const snap = await db.ref(`matchmakingQueue/${mode}`).get();
    queuesByMode[mode] = snap.val() || {};
  }

  const counts = countQueue(queuesByMode, Date.now());
  const committed = await countCommitments();
  await getFirestore().collection("stats").doc("presence").set({
    ...counts,
    ...committed,
    updatedAt: FieldValue.serverTimestamp(),
  });
  return {...counts, ...committed};
}

/**
 * How many people have said they're in for tonight's window.
 *
 * Piggybacks on this once-a-minute job rather than getting its own,
 * because it publishes to the same document clients already listen to -
 * so a second number costs no extra write, no extra listener, and no extra
 * scheduled job. It also inherits the same honesty rule as the live count:
 * the figure is only worth showing while it's true.
 *
 * Uses an aggregation count() rather than reading the matching documents,
 * which stays cheap as the userbase grows - the whole point is that this
 * number gets larger.
 *
 * Best-effort: a failure here returns nothing rather than throwing, so a
 * broken commitment count never takes the live online count down with it.
 */
async function countCommitments(now = new Date()) {
  try {
    const db = getFirestore();
    const configSnap = await db.collection("config").doc("eventWindow").get();
    const config = readEventWindowConfig(configSnap.data());
    if (!config.enabled) return {};
    const dayKey = upcomingWindowDayKey(now, config);
    const agg = await db.collection("users")
        .where("eventCommitmentDayKey", "==", dayKey)
        .count().get();
    return {committedTonight: agg.data().count, commitmentDayKey: dayKey};
  } catch (e) {
    console.error("countCommitments failed:", e.message);
    return {};
  }
}

module.exports = {publishOnlineCount, countQueue, countCommitments, PRESENCE_STALE_MS};

File: lib/core/services/presence.dart
import 'package:cloud_firestore/cloud_firestore.dart';

/// How many people are queueing or mid-match right now, published once a
/// minute by the `publishOnlineCount` Cloud Function.
///
/// The count's only value is being true - it exists to answer "will anyone
/// be there?", and an inflated number sends someone into an empty queue and
/// teaches them the app is dead. So this errs hard towards showing nothing:
/// a missing document, an unreadable one, or a count published too long ago
/// all render as "unknown" rather than as a guess.
class OnlineCount {
  const OnlineCount({
    required this.waiting,
    required this.matched,
    required this.total,
    required this.updatedAt,
  });

  final int waiting;
  final int matched;
  final int total;
  final DateTime? updatedAt;

  /// The publisher runs every minute, so anything much older than that
  /// means it has stopped running. Showing a frozen number from an hour ago
  /// would be worse than showing none - it is the precise failure this
  /// counter must not have.
  static const Duration _maxAge = Duration(minutes: 5);

  bool get isFresh {
    final at = updatedAt;
    if (at == null) return false;
    return DateTime.now().toUtc().difference(at.toUtc()) <= _maxAge;
  }

  static OnlineCount? fromMap(Map<String, dynamic>? data) {
    if (data == null) return null;
    final updatedAt = data['updatedAt'];
    return OnlineCount(
      waiting: (data['waiting'] as num?)?.toInt() ?? 0,
      matched: (data['matched'] as num?)?.toInt() ?? 0,
      total: (data['total'] as num?)?.toInt() ?? 0,
      updatedAt: updatedAt is Timestamp ? updatedAt.toDate() : null,
    );
  }

  /// Phrased as people rather than as a bare number, because "3 roasters
  /// online" reads as a room and "3" reads as a statistic.
  String get label {
    if (total == 1) return '1 roaster online now';
    return '$total roasters online now';
  }
}

Stream<OnlineCount?> onlineCountStream() {
  return FirebaseFirestore.instance
      .collection('stats')
      .doc('presence')
      .snapshots()
      .map((snap) => OnlineCount.fromMap(snap.data()))
      .handleError((_) => null);
}
